Add edge-style property

zakjan opened the pull request:
Adds a new `edge-style` property, may be `line` (default), or `triangle`.

![Screen Shot 2021-06-28 at 9 47 56](https://user-images.githubusercontent.com/173585/123599505-f3ab9600-d7f5-11eb-8f8d-fd4b8e1edce2.png)

Closes #2883

zakjan commented:
This is a PR

maxkfranz commented:
Looks good.  Since this is conceptually the same as `curve-style`, could this be amended to use the existing property?  It would be simpler for library consumers and less error-prone in the code in the long term.

zakjan commented:
Initially I thought so, but it's incompatible with most of line styles. Is it ok?

maxkfranz commented:
Isn't it incompatible with other properties in any case (e.g. control points)?  My line of thinking is that one property where the value seems to fit sensibly is simpler than two.  In any case, incompatibilities should be noted where they wouldn't be obvious (e.g. the edge offset distance from the node probably works, but some others may not in subtle ways).

If a property is already listed inclusively (e.g. "the following properties only work for `curve-style: foo | bar`), then those parts are fine as-is.  In cases where it may be ambiguous, we should try to make things clear for users.

zakjan commented:
Updated, it's `curve-style: straight-triangle` now

maxkfranz commented:
Great.  I've tested out your changes a bit.  Here are some questions:

(1) I added an `ef` edge in the debug demo to try it out.  Could you add an edge like that to your PR so that it could be more easily tested in future?

```
.selector('#ef')
        .style({ // e.g.
          'curve-style': 'straight-triangle',
          'width': 6

          // then you can play around with different styles to make sure they work in combination, e.g.:
          'source-endpoint': 'outside-to-node-or-label',
          'source-distance-from-node': 10,
        })
```

(2) The arrows seem to be shown even though you say in the documentation that they're unsupported.  Should the documentation be changed, or should the arrows be disabled in the code?

(3) The documentation doesn't explicitly say which direction the triangle goes, i.e. which is the small, pointed side (target) and which is the big, flat side (source).  You hint about the direction a bit, but it could be unclear what it means precisely to a novice of the library.  Could this be amended?

(4) Which [ edge endpoint](https://js.cytoscape.org/#style/edge-endpoints) properties are compatible with the new edge type?  If it's intended to support all of them, have you tested those cases?  Everything must be compatible with existing properties, or where not possible the documentation should explicitly note incompatibilities.

(5) Are there any [line](https://js.cytoscape.org/#style/edge-line) properties that are incompatible (e.g. properties that make sense to be compatible like `line-cap` should be tested and confirmed working)?  Which combinations have you tested?  Please note incompatibilities in the docs.

zakjan commented:
Thanks for highlighting possible conflicts with other existing features.

1) There is a testing edge in `edge-types` demo. Is it good enough, or do you mean another demo?
2) You're right, arrows work well. I removed the note about arrows not supported.
3) Updated docs.
4) Tested all edge endpoint properties. All values work well.
5) Tested all line properties. Solid fills including gradients work well. `line-style`, `line-cap`, `line-dash-pattern`, `line-dash-offset` are not supported. Updated docs.

There are no changes in the code, only in the docs. Does it cover all your concerns?

maxkfranz commented on documentation/md/style.md:
```diff
@@ -420,6 +420,15 @@ For straight line edges (`curve-style: straight`, [demo](demos/edge-types)):
+
+For straight triangle edges (`curve-style: straight-triangle`, [demo](demos/edge-types)):
+
+A straight triangle edge (`curve-style: straight-triangle`) is drawn as a single straight isosceles triangle in the direction from the source to the target.
```
Which side is source and which is target?  This should be described explicitly (pointy side vs. flat side, or small side versus large side, etc.)

maxkfranz commented on documentation/md/style.md:
```diff
@@ -420,6 +420,15 @@ For straight line edges (`curve-style: straight`, [demo](demos/edge-types)):
+
+A straight triangle edge (`curve-style: straight-triangle`) is drawn as a single straight isosceles triangle in the direction from the source to the target.
+
+Property `width` defines line width at the triangle base, line width is `0` at the triangle apex.  Properties `line-style`, `line-cap`, `line-dash-pattern`, `line-dash-offset` are not supported.
```
> line width is `0` at the triangle apex

It's probably more intuitive to describe this as a point

zakjan replied:
Updated, moved to the first paragraph

maxkfranz reviewed (commented):
> There is a testing edge in edge-types demo. Is it good enough, or do you mean another demo?

All edge types should be included in the debug demo, as the debug demo is used by mostly everyone when making contributions.  This makes it very likely that someone working on the project would notice if there were any conflicts with a particular edge type and their work.

> You're right, arrows work well. I removed the note about arrows not supported.

Great

> Updated docs.

See comments

> Tested all edge endpoint properties. All values work well.

OK

> Tested all line properties. Solid fills including gradients work well. line-style, line-cap, line-dash-pattern, line-dash-offset are not supported. Updated docs.

What if a library consumer is looking at the individual line properties rather than the triangle curve style section?  Should these conflicts also be mentioned alongside those properties themselves?

maxkfranz commented:
> Re demo, I agree with you, but as I mentioned, the new edge type is already present in the demo. This is the screenshot. Is this good enough? If not, I'm afraid that I'm still not sure what do you mean.
> 

Take a look here: https://github.com/cytoscape/cytoscape.js/tree/unstable/debug

zakjan commented:
Oh, thanks, I didn't know about this separate demo. Added.
